lib/std/smartptrs.nim

```diff
@@ -0,0 +1,130 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2021 Nim contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+
+## C++11 like smart pointers. They always use the shared allocator.
+import std/isolation
+
+type
+  UniquePtr*[T] = object
+    ## Non copyable pointer to a value of type `T` with exclusive ownership.
+    val: ptr T
+
+proc `=destroy`*[T](p: var UniquePtr[T]) =
+  if p.val != nil:
+    `=destroy`(p.val[])
+    deallocShared(p.val)
+
+proc `=copy`*[T](dest: var UniquePtr[T], src: UniquePtr[T]) {.error.}
+  ## The copy operation is disallowed for `UniquePtr`, it
+  ## can only be moved.
+
+proc newUniquePtr[T](val: sink Isolated[T]): UniquePtr[T] {.nodestroy.} =
```

Comment:
how about naming it uniquePtr instead? new is a misnomer since it's still an object type underneath, and init also doesn't reflect ptr-like semantics

ditto for other constructors

Reply:
what's the value on that? it's not unheard of to use new for objects placed on the heap in Nim. Moreover this api is already from 2019, pleople already use it, its not like experimental.

Reply:
`UniquePtr` itself is not placed on the heap

Reply:
...and so are pointers.